# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [2]:
def add_two(x):
    return x+2
print(add_two(3))


5


In [ ]:
lambda x: x+2
(lambda x )

#### Saving to a Variable

In [3]:
f =lambda x, y: x+y+2


<function __main__.<lambda>(x)>

#### Multiple Inputs

In [4]:
(lambda x, y, z:x*y*z)(3,5,8)

120

#### Passing a Lambda into a Function

In [ ]:
def do_math(func, num):
    return func(num)
do_math: lambda x: x*x, 4

#### Returning a Lambda from a Function

In [6]:
def my_func():
    return "jello"
print(my_func)
print(type(my_func()))

<function my_func at 0x7fa70c10f310>
<class 'str'>


In [18]:
def return_func():
    def squared(num):
        return num*num
  
    return squared
print(return_func()(3))

9


In [19]:
def return_lamb(b,c):
    return lambda x,y: (b+c)-(x+y)
return_lamb(3,2)(2,3)

0

#### If Statements within Lambdas

In [21]:
#Lambda x : True if (condition) else false
f = lambda num: num*2 if num > 10 else num+2
f(12)

24

#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [22]:
f = lambda x: x**3
f(3)

27

In [23]:
#get good at reading Documentation because this is your lifelinealso
help(sorted)

Help on built-in function sorted in module builtins:

sorted(iterable, /, *, key=None, reverse=False)
    Return a new list containing all items from the iterable in ascending order.
    
    A custom key function can be supplied to customize the sort order, and the
    reverse flag can be set to request the result in descending order.



In [ ]:
#Using Lambdas with sort to sort a list of dictionary by a key from highest to smallest


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [24]:
# map(func,iterable(list,dict,tuple,set,etc...))
# Normally the usage of map happens with a pre-defined function -- but can use lambdas as well
help(map)

Help on class map in module builtins:

class map(object)
 |  map(func, *iterables) --> map object
 |  
 |  Make an iterator that computes the function using arguments from
 |  each of the iterables.  Stops when the shortest iterable is exhausted.
 |  
 |  Methods defined here:
 |  
 |  __getattribute__(self, name, /)
 |      Return getattr(self, name).
 |  
 |  __iter__(self, /)
 |      Implement iter(self).
 |  
 |  __next__(self, /)
 |      Implement next(self).
 |  
 |  __reduce__(...)
 |      Return state information for pickling.
 |  
 |  ----------------------------------------------------------------------
 |  Static methods defined here:
 |  
 |  __new__(*args, **kwargs) from builtins.type
 |      Create and return a new object.  See help(type) for accurate signature.



In [25]:
def squared(num, num2):
    if num < 10 and num2 <10:
        return num**2, num2**2
    else:
        return num, num2
    
numbers = [4,10,3,2,6]
more_nums = [4,1,20,3,15]

squared_nums_map = map(squared,numbers,more_nums)
print(squared_nums_map)
print(list(squared_nums_map))

#### Using Lambda's with Map

In [ ]:
# map(Lambda x+2,list)
# using Lambda in map happens in line (or in one line)
# list(map(lambda x: x*2,nums))
numbers = [4, 10, 3, 2, 6 ]
more_nums = [4, 1, 20, 3, 15]

squared_num2= list(map((lambda x,y:(x**2, y**2))))

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [ ]:
nums = [4,2,6,8]
lambda newnums: map(,)

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [26]:
list(filter(None, [True, True, False, False]))


[True, True]

#### Using Lambda's with Filter()

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [27]:
from statistics import mean

list1 = [4,5,6,3,7,2,8]
new_list = list(filter((lambda num:True if num<mean(list1) else False), list1))
print(new_list)

[4, 3, 2]


## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [30]:
from functools import reduce
# reduce(function,iterable)

list_1= [2, 4, 6, 8, 10]
def subtract_nums(num1, num2):
    return num1 - num2 
result = reduce(subtract_nums, list_1)
result

-26

#### Using Lambda's with Reduce()

In [ ]:
l_3 = [1,10,2,30]
result_lamb = reduce((lambda num1, num2: num1-num2),l_3)
result_lamb

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [31]:
l_3 = [1,10,2,30]
result_lamb = reduce((lambda num1, num2: num1*num2),l_3)
result_lamb

600

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

In [39]:
def recursive(i):
    print("hi")
    if i == 4:
        return
    else:
        i+=1
        recursive(i)
recursive(0)    

hi
hi
hi
hi
hi


#### Implementing a Base Case

In [38]:
def add_nums(num):
    #set base case
    if num<=1:
        print("add_nums(1)=1")
        return num
    else:
        print(f"add_nums({num})= {num} + add_nums({num-1})")
        x = num- 1
        return num + add_nums(x)

#### Writing a Factorial Function

In [ ]:
# 5! = 5*4*3*2*1

def factorial(num):
    if num<= 1:
        print('factorial(1)= 1')

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

# Exercises

### Exercise #1 <br>
starter for ex 1
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [24]:
places = ["","Argentina", "", "San Diego","","","","Boston","New York"]
place=list(filter(None,places))
place

['Argentina', 'San Diego', 'Boston', 'New York']

### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>
stater for ex #2
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [25]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]
author.sort(key=lambda x: x, reverse=True)
print(author)

['Victor aNisimov', 'Joel Carter', 'Gary A.J. Bernstein', 'David hassELHOFF', 'Andrew P. Garfield']


### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]
`

In [40]:
# F = (9/5)*C + 32 

places = [('Nashua',32),("Boston",12),("Los Angelos",44),("Miami",29)]
temp = lambda x: (x[0], (9/5) * x[1] + 32)
list(map(temp,places))

[('Nashua', 89.6), ('Boston', 53.6), ('Los Angelos', 111.2), ('Miami', 84.2)]

### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`

In [11]:
 def fib(n):
        if n ==1:
            return 1
        elif n == 2:
            return 1
        elif n > 2:
            return fib(n-1)+ fib(n-2)
            
for n in range(1,6):
    print(n, ':', fib(n))


0 : None
1 : 1
2 : 1
3 : 2
4 : 3
5 : 5
